File: python/day3/day3.py
from typing import List
from dataclasses import dataclass
from functools import reduce


def load_input_text() -> List[str]:
    with open("python/day3/day-3-input.txt") as f:
        input = f.readlines()
    return [i.strip() for i in input]


@dataclass
class Point:
    x: int
    y: int

@dataclass
class Move:
    x: int
    y: int

@dataclass
class LocalGeology:
    layout: List[str]
    position: Point
    movement: Move

    def move(self):
        self.position = Point(self.position.x + self.movement.x, self.position.y + self.movement.y)
        self.position.x = self.position.x % len(self.layout[0])

        return self
    
    def is_on_tree(self):
        if self.layout[self.position.y][self.position.x] == "#":
            return True
        else:
            return False

    def has_reached_end(self):
        if self.position.y >= len(self.layout) - 1:
            return True
        else:
            return False

    def slide_to_end():
        pass



if __name__=="__main__":
    movement_choices = [(1,1),(3,1),(5,1),(7,1),(1,2)]
    movement_tree_sum = []

    for choice in movement_choices:
        start = Point(0,0)
        r = LocalGeology(load_input_text(), start, Move(*choice))
        sum = 0
        while not r.has_reached_end():
            r = r.move()
            sum += r.is_on_tree()
        movement_tree_sum.append(sum)
        
    
    print(reduce(lambda a,b: a*b, movement_tree_sum))

File: rust/aoc/src/day3.rs
use std::fs;

fn load_input_text() -> Vec<String> {
    // read the file and unwrap to panic if the file isn't readable
    let contents = fs::read_to_string("day-3-input.txt")
        .unwrap();

    let layout: Vec<String> = 
        contents.trim().split('\n')
        .map(String::from) // convert from reference &str to String value
        .collect();
        
    layout
}

#[derive(Debug)]
struct Point {
    x: usize,
    y: usize
}

#[derive(Debug)]
struct Move {
    x: usize,
    y: usize
}

#[derive(Debug)]
struct LocalGeology {
    layout: Vec<String>,
    position: Point,
    movement: Move
}

impl LocalGeology {
    fn move_position(mut self) -> LocalGeology {
        self.position = Point{x: self.position.x + self.movement.x, y: self.position.y + self.movement.y};
        self.position.x = self.position.x % self.layout[0].chars().count();

        self
    }

    fn is_on_tree(&self) -> bool {
        let row: &str = &self.layout[self.position.y];
        let cell: String = row.chars().nth(self.position.x).unwrap().to_string();
        if cell == "#" {
            return true;
        }
        return false;
    }

    fn has_reached_end(&self) -> bool {
        if self.position.y >= (self.layout.len()-1) {
            return true
        }
        return false;
    }

    fn slide_to_end(&self) -> () {
        ()
    }

}


pub fn solve() -> u32 {
    let movement_choices = [(1,1),(3,1),(5,1),(7,1),(1,2)];
    let mut movement_tree_sum: Vec<u32> = Vec::new();


    for choice in movement_choices.iter() {
        let start_position = Point{x:0, y:0};
        let mut r = LocalGeology{
            layout: load_input_text(),
            position: start_position,
            movement: Move{x: choice.0, y: choice.1}
        };
        let mut sum = 0;

        loop {
            if r.has_reached_end() == true {
                break;
            }
            r = r.move_position();
            sum += r.is_on_tree() as u32
        }
        movement_tree_sum.push(sum);
        println!("Done {:?}! Sum is {}", choice, sum);
    }

    let total = movement_tree_sum.iter().fold(1, |acc,x| acc * x);
    total
}